### PR TITLE
Feature/holder endpoint

### PR DESCRIPTION
**Pull request by VargaElod23**

### CODE REVIEW

File: internal/storage/types.go [hunk 1]
```diff
@@ -30,6 +30,11 @@ type AddressStats struct {
 	mutex   sync.RWMutex `rlp:"-"`
 }
 
+type Account struct {
+	models.Account
+	mutex sync.RWMutex `rlp:"-"`
+}
+
```
Comment (kstdl) [inline internal/storage/types.go:37]: I think its better to have something like `Balance *big.Int` here to not parse it on every modification. And convert to a model with string values just before you will save it to db

File: internal/utils/events.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		err := contractABI.UnpackIntoInterface(&event, "CommissionRewardsClaimed", data)
+		if err != nil {
+			return nil, err
+		}
+		return &event, nil
+
+	default:
+		return nil, fmt.Errorf("unknown event topic")
+	}
+}
```
Comment (kstdl) [inline internal/utils/events.go:69]: You should be checking somewhere if this events are from our DPOS contract

File: internal/indexer/transactions.go [hunk 3]
```diff
@@ -40,6 +43,47 @@ func (bc *blockContext) processTransactions(trxHashes *[]string) (err error) {
 			Data: trx.ExtractLogs(),
 		}
 		bc.batch.AddToBatchSingleKey(logs, trx_model.Hash)
+		tp.Go(utils.MakeTask(bc.updateHolderBalances, trx, &err).Run)
+	}
+	return
+}
+
+func (bc *blockContext) updateHolderBalances(trx chain.Transaction) (err error) {
+
```
Comment (kstdl) [inline internal/indexer/transactions.go:52]: Why do you have empty lines on every method start? :)

File: internal/storage/types.go (hunk 1 above)
Comment (VargaElod23) [reply internal/storage/types.go:37]: Right, I'll change this

File: models/models.gen.go [hunk 4]
```diff
@@ -13,6 +13,12 @@ const (
 	Transfer                 TransactionType = 0
 )
 
+// Account defines model for Account.
+type Account struct {
+	Address *string `json:"address,omitempty"`
+	Balance *string `json:"balance,omitempty"`
```
[the remark this answers is not in the document]
Comment (VargaElod23) [reply models/models.gen.go:19]: Not sure why the generation creates them as pointers as both are simple strings

File: internal/indexer/transactions.go (hunk 3 above)
Comment (VargaElod23) [reply internal/indexer/transactions.go:52]: Old Java things 😆 sorry, will format

File: internal/storage/pebble/storage.go [hunk 5]
```diff
@@ -202,6 +204,51 @@ func (s *Storage) GetFinalizationData() *storage.FinalizationData {
...
+func (s *Storage) GetAccounts() *map[string]*models.Account {
+	ptr := make(map[string]*models.Account)
+	err := s.getFromDB(&ptr, getKey(getPrefix(ptr), "0x0", 0))
+	if err != nil && err != pebble.ErrNotFound {
+		log.Fatal("GetAccounts ", err)
+	}
+	return &ptr
+}
+
+func (s *Storage) HolderModelsToStorage(ptr *map[string]*models.Account) *map[string]*storage.Account {
```
Comment (kstdl) [inline internal/storage/pebble/storage.go:236]: Why is this in storage?

File: internal/storage/storage.go [hunk 6]
```diff
@@ -65,3 +70,45 @@ func GetObjectsPage[T Paginated](s Storage, address string, from, count uint64)
 	})
 	return
 }
+
+func GetAccountsPage[T Paginated](s Storage, from, count uint64) (ret []T, pagination *models.PaginatedResponse) {
+
```
Comment (kstdl) [inline internal/storage/storage.go:75]: formatting :)

File: internal/storage/storage.go [hunk 7]
```diff
@@ -65,3 +70,45 @@ func GetObjectsPage[T Paginated](s Storage, address string, from, count uint64)
...
+		acc := account.ToModel()
+		t := reflect.TypeOf(acc)
+		value := reflect.New(t).Elem()
+
+		// Copy values from acc to the new instance
+		value.Set(reflect.ValueOf(acc))
+
+		// Append the new instance to the ret slice
+		ret = append(ret, value.Interface().(T))
+		account.Mutex.RUnlock()
```
Comment (kstdl) [inline internal/storage/storage.go:109]: Whole this thing could be replaced with this thing, or am I missing something? 
```ret = append(ret, account.ToModel())```

File: internal/storage/storage.go [hunk 8]
```diff
@@ -65,3 +70,45 @@ func GetObjectsPage[T Paginated](s Storage, address string, from, count uint64)
 	})
 	return
 }
+
+func GetAccountsPage[T Paginated](s Storage, from, count uint64) (ret []T, pagination *models.PaginatedResponse) {
+
+	pagination = new(models.PaginatedResponse)
+	pagination.Start = from
+	accountMap := s.GetAccounts()
+	storageAccountMap := s.HolderModelsToStorage(accountMap)
```
Comment (kstdl) [inline internal/storage/storage.go:79]: Better to convert it only for elements that you are returning in the cycle below

File: internal/storage/types.go [hunk 9]
```diff
@@ -13,7 +13,7 @@ import (
 type TotalSupply = big.Int
 
 type Paginated interface {
-	models.Transaction | models.Dag | models.Pbft
+	models.Account | models.Transaction | models.Dag | models.Pbft
```
Comment (kstdl) [inline internal/storage/types.go:16]: You added new generic method for the Account. So no need to add it here as you aren't using it in `GetObjectsPage` or other methods

File: internal/storage/types.go [hunk 10]
```diff
@@ -30,6 +30,12 @@ type AddressStats struct {
 	mutex   sync.RWMutex `rlp:"-"`
 }
 
+type Account struct {
+	Address string       `json:"address"`
+	Balance *big.Int     `json:"balance"`
+	Mutex   sync.RWMutex `rlp:"-"`
```
Comment (kstdl) [inline internal/storage/types.go:36]: If I am not mistaken you aren't changing this concurrently, so you don't need mutex here

**Pull request closed without merging**